### PR TITLE
cmake: c++ exceptions linking support

### PR DESCRIPTION
**Pull request by tejlmand**
Fixes part of: #32448

This commit updates the CMake CMAKE_CXX_LINK_EXECUTABLE to include
crtbegin.o and crtend.o at the right locations when linking with gcc.

It also updates linker scripts to ensure proper location of the
exception header frame sections.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>
Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

--------
Update:
The following platforms are fixed with this PR:
qemu_x86
qemu_riscv32
qemu_xtensa

The following platforms still fails:
qemu_x86_64 (fails with `exit`)
qemu_cortex_a53 (fails with `exit`)
qemu_riscv64 (fails with `exit`)
qemu_nios2 (hangs)
qemu_leon3 (fails with trap_instruction)

### CODE REVIEW

**Comment by tejlmand** [comment]
@galak I included your linker script modifications from this: #34357

**Comment by dcpleung** [comment]
Could you refactor the linker script addition into a file, so it can be done via #include?

**Comment by tejlmand** [comment]
> Could you refactor the linker script addition into a file, so it can be done via #include?

Done, placed it in cplusplus-ram.ld where we already have linker code related to exception tables.

**Review by galak — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Review by stephanosio — COMMENTED** [review]
This claims to fix #32448, but there are some platforms that still fail.

Just noting that #32448 should stay open to keep track of them ... (see https://github.com/zephyrproject-rtos/zephyr/issues/32448#issuecomment-797164146).

**Review by stephanosio — CHANGES_REQUESTED** [review]
The following should be updated to test the platforms that work after the changes in this PR:
https://github.com/zephyrproject-rtos/zephyr/blob/7b7ed2baac4f954ead85b542444cd9b143e5ddba/tests/subsys/cpp/libcxx/testcase.yaml#L11-L19

**Comment by tejlmand** [comment]
> This claims to fix #32448, but there are some platforms that still fail.

Looking into this, the `CPP_TEST` provided in #32448 passes on the platforms: qemu_xtensa, qemu_riscv32, qemu_cortex_m3 and qemu_x86 works.
As example:
```
$ west build -b qemu_xtensa --pristine -t run
-- west build: making build dir /projects/github/ncs/zephyr/tests/CPP_TEST/build pristine
...
*** Booting Zephyr OS build zephyr-v2.5.0-4098-ge2501fd08ef3  ***
Running test suite framework_tests
===================================================================
START - test_assert
 PASS - test_assert in 0.1 seconds
===================================================================
START - test_cexception
 PASS - test_cexception in 0.2 seconds
===================================================================
Test suite framework_tests succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

also some archs passes on `tests/subsys/cpp/libcxx` when built and executed manually, like qemu_xtensa:
```
$  west build -b qemu_xtensa --pristine -t run -- -DCONFIG_EXCEPTIONS=y
-- west build: making build dir /projects/github/ncs/zephyr/tests/subsys/cpp/libcxx/build pristine
...
*** Booting Zephyr OS build zephyr-v2.5.0-4098-ge2501fd08ef3  ***
version 201703
Running test suite libcxx_tests
===================================================================
START - test_array
 PASS - test_array in 0.1 seconds
===================================================================
START - test_vector
 PASS - test_vector in 0.1 seconds
===================================================================
START - test_make_unique
 PASS - test_make_unique in 0.1 seconds
===================================================================
START - test_exception
 SKIP - test_exception in 0.1 seconds
===================================================================
Test suite libcxx_tests succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
qemu-system-xtensa: terminating on signal 2
```

but  fails when using twister:
```
$ ./scripts/twister -v -p qemu_xtensa -N  --inline-logs -s  tests/subsys/cpp/libcxx/cpp.libcxx.exceptions
...
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.

INFO    - /projects/github/ncs/zephyr/twister-out/qemu_xtensa/tests/subsys/cpp/libcxx/cpp.libcxx.exceptions/build.log

INFO    - 0 of 1 test configurations passed (0.00%), 1 failed, 0 skipped with 0 warnings in 5.36 seconds
INFO    - In total 4 test cases were executed, 0 skipped on 1 out of total 372 platforms (0.27%)
INFO    - 0 test configurations executed on platforms, 1 test configurations were only built.
```

other archs (riscv32) fails on tests/subsys/cpp but not CPP_TEST.

Marked `DNM` to allow further investigations.

File: cmake/linker/ld/target.cmake [hunk 1]
```diff
@@ -4,6 +4,15 @@ find_program(CMAKE_LINKER     ${CROSS_COMPILE}ld      PATHS ${TOOLCHAIN_HOME} NO
 
 set_ifndef(LINKERFLAGPREFIX -Wl)
 
+if(CONFIG_EXCEPTIONS)
+  # When building with C++ Exceptions, it is important that crtbegin and crtend
+  # are linked at specific locations.
+  # The location is so important that we cannot let this be controlled by normal
+  # link libraries, instead we must control the link command specifically as
+  # part of toolchain.
```
[the remark this answers is not in the document]
Comment (galak) [reply cmake/linker/ld/target.cmake:12]: For GCC this has normally be done via a spec file.  @andyross Have you seen anywhere that this is done via linker script?

[the remark this answers is not in the document]
Comment (tejlmand) [reply cmake/linker/ld/target.cmake:12]: how will you tell the linker which objects (crtbegin.o and crtend.o) to link in a linker file ?
Not to mention that the position of those objects on the linker invocation is important.

This is why this is to be considered a toolchain (linker) setting.

**Comment by tejlmand** [comment]
Updated description text with list of platform that are fixed with this PR, as it stands.
The following platforms are fixed with this PR:
qemu_x86
qemu_riscv32
qemu_xtensa

The following platforms still fails:
qemu_x86_64 (fails with `exit`)
qemu_cortex_a53 (fails with `exit`)
qemu_riscv64 (fails with `exit`)
qemu_nios2 (hangs)
qemu_leon3 (fails with trap_instruction)

**Comment by tejlmand** [comment]
> The following should be updated to test the platforms that work after the changes in this PR:

Done.

**Review by stephanosio — APPROVED** [review]

**Pull request merged**